### ISSUE TITLE
Should module function type annotations always be required?

### ISSUE THREAD

**Issue by lpil** [gleam-lang/suggestions#48]
Currently we are like Elm, Haskell, and OCaml in that type annotations are not required, but they can be given.

```rust
pub fn main(name) { ... }

// or

pub fn main(name: String) { ... }

// or

pub fn main(name: String) -> Int { ... }
```

In real-world code written in these languages type annotations are almost always given for top level functions, and in Elm the compiler will complain if you do not supply them.

Other popular languages with static type systems always require type annotations for top level functions. i.e. Rust, Scala, Kotlin, Go.

With annotated function arguments we can provide easier to understand messages when there is a type error.
It would also make it trivial to introduce mutual recusion, remove the need for functions to be defined in the order they are used, and enable some fancier type features in addition.

Lastly I think having more certainty in the types when being inferred would open the door to some overloading or implicit typing features that would otherwise introduce ambiguity.

The cost of requiring these annotations is that there is more boilerplate for the programmer, though not an unusual amount for a statically typed language.

I'm interested in hearing what other people think about this question.

**Comment by Janiczek** [comment]
> There are two main advantages to always requiring module function annotations that I have not seen other people mention in this discussion yet:
>
>    Compilation will be faster because there is less to infer.

@Qqwy I'm unsure if you mean it like "user gave us this type, let's skip inferring the actual body to see if they're right" - that shouldn't be done IMHO: your type annotation, if given, should still be checked against the real type, and error raised if you were wrong.

**Comment by lpil** [comment]
That's right, languages that have mandatory type annotations still perform full type inference to ensure the annotations are correct.

**Comment by lpil** [comment]
I'm going to close this as I believe the main issues have been resolved in other ways now. Thanks all!

**Comment by Qqwy** [comment]
For posterity, I meant that when the user's type annotation already fills in certain type variables in the signature, then this narrows down the unification that is required to test whether the real type and the provided type match. ☺️

On September 12, 2020 11:14:33 PM UTC, Martin Janiczek <notifications@github.com> wrote:
>> There are two main advantages to always requiring module function
>annotations that I have not seen other people mention in this
>discussion yet:
>>
>>    Compilation will be faster because there is less to infer.
>
>@Qqwy I'm unsure if you mean it like "user gave us this type, let's
>skip inferring the actual body to see if they're right" - that
>shouldn't be done IMHO: your type annotation, if given, should still be
>checked against the real type, and error raised if you were wrong.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/gleam-lang/suggestions/issues/48#issuecomment-691569963

-- 
Met vriendelijke groet / Sincerely, 
~W-M
